File: windows-app/Plantoir.Tests/SiteHealthContractTests.cs
using System.Text.Json.Nodes;
using Plantoir.Core.Models;

namespace Plantoir.Tests;

/// <summary>
/// The site-health contract, run against this app's own parser.
///
/// Every case here is deserialised from
/// <c>contracts/shared-rules.json</c> -> <c>siteHealth</c>. Nothing is
/// retyped: a sentence written into a test file is the copy that keeps
/// passing after the product's words change, which is the whole reason the
/// contract exists.
/// </summary>
public class SiteHealthContractTests
{
    private static JsonNode SiteHealth =>
        ContractLoader.LoadJson("shared-rules.json")["siteHealth"]!;

    /// <summary>The marker this app looks for is the one the build prints.</summary>
    [Fact]
    public void Marker_MatchesContract()
    {
        string prefix = SiteHealth["marker"]!["prefix"]!.ToString();
        Assert.Equal(SiteHealthFinding.Marker, prefix);
    }

    public static TheoryData<string> MarkerExamples()
    {
        var data = new TheoryData<string>();
        foreach (var example in SiteHealth["marker"]!["examples"]!.AsArray())
            data.Add(example!.ToString());
        return data;
    }

    /// <summary>
    /// The contract's own example lines parse, carrying every field. These are
    /// real lines copied from a build, so a parser that passes here handles
    /// what the launcher actually emits.
    /// </summary>
    [Theory]
    [MemberData(nameof(MarkerExamples))]
    public void ContractExampleLines_Parse(string line)
    {
        var finding = SiteHealthFinding.Parse(line);
        Assert.NotNull(finding);
        Assert.NotEmpty(finding!.Name);
        Assert.NotEmpty(finding.Sentence);
        Assert.NotEmpty(finding.Detail);
        Assert.NotEmpty(finding.Course);
        Assert.True(finding.Section > 0);
    }

    public static TheoryData<string, bool> ContractChecks()
    {
        var data = new TheoryData<string, bool>();
        foreach (var check in SiteHealth["checks"]!.AsArray())
            data.Add(check!["name"]!.ToString(), check["fixable"]!.GetValue<bool>());
        return data;
    }

    /// <summary>
    /// Every check in the contract survives a round trip through the parser
    /// with its own sentence and detail intact -- including
    /// <c>noGradedFolders</c>, which is the check this piece added. A check
    /// added to the contract on either platform arrives here as a new case
    /// without anybody editing this file.
    /// </summary>
    [Theory]
    [MemberData(nameof(ContractChecks))]
    public void EveryContractCheck_SurvivesTheLine(string name, bool fixable)
    {
        var check = SiteHealth["checks"]!.AsArray()
            .First(c => c!["name"]!.ToString() == name)!;
        string sentence = check["sentence"]!.ToString()
            .Replace("{course}", "ICS3U").Replace("{section}", "1");
        string detail = check["detail"]!.ToString();

        var payload = new JsonObject
        {
            ["name"] = name,
            ["sentence"] = sentence,
            ["detail"] = detail,
            ["fixable"] = fixable,
            ["course"] = "ICS3U",
            ["section"] = 1,
        };

        var finding = SiteHealthFinding.Parse($"{SiteHealthFinding.Marker} {payload.ToJsonString()}");

        Assert.NotNull(finding);
        Assert.Equal(name, finding!.Name);
        Assert.Equal(sentence, finding.Sentence);
        Assert.Equal(detail, finding.Detail);
        Assert.Equal(fixable, finding.Fixable);
        Assert.Equal("ICS3U", finding.Course);
        Assert.Equal(1, finding.Section);
    }

    /// <summary>
    /// A Fix button is offered for exactly the checks the contract says it may
    /// be offered for -- decided from the NAME, not from the fixable flag.
    /// </summary>
    [Theory]
    [MemberData(nameof(ContractChecks))]
    public void RepairIsOfferedForExactlyTheContractsChecks(string name, bool fixable)
    {
        var repair = SiteHealth["repair"]!;
        var offered = repair["offered"]!["checks"]!.AsArray()
            .Select(c => c!.ToString()).ToHashSet();
        var neverOffered = repair["neverOffered"]!["checks"]!.AsArray()
            .Select(c => c!.ToString()).ToHashSet();

        Assert.True(offered.Contains(name) || neverOffered.Contains(name),
            $"The contract classifies every check as offered or never offered; {name} is in neither.");

        var finding = new SiteHealthFinding(name, "s", "d", fixable, "ICS3U", 1);
        Assert.Equal(offered.Contains(name), finding.CanBeRepaired);
    }

    /// <summary>
    /// <c>noGradedFolders</c> in particular is never repairable. Named on its
    /// own because it is the check this piece added, and because an existence
    /// fix would not assign marks to any page -- the Fix button would report
    /// success having changed nothing a teacher wanted changed.
    /// </summary>
    [Fact]
    public void NoGradedFolders_IsNeverOfferedARepair()
    {
        var neverOffered = SiteHealth["repair"]!["neverOffered"]!["checks"]!.AsArray()
            .Select(c => c!.ToString()).ToList();
        Assert.Contains("noGradedFolders", neverOffered);

        var finding = new SiteHealthFinding("noGradedFolders", "s", "d", false, "ICS3U", 1);
        Assert.False(finding.CanBeRepaired);
    }

    /// <summary>
    /// The code's repairable list holds nothing the contract has not approved.
    /// The theory above only visits names the contract knows; this catches a
    /// name added to the code and to neither list.
    /// </summary>
    [Fact]
    public void TheCodeOffersNoRepairTheContractDoesNot()
    {
        var repair = SiteHealth["repair"]!;
        var offered = repair["offered"]!["checks"]!.AsArray()
            .Select(c => c!.ToString()).ToHashSet();
        var neverOffered = repair["neverOffered"]!["checks"]!.AsArray()
            .Select(c => c!.ToString()).ToHashSet();

        Assert.Equal(offered, SiteHealthFinding.RepairableChecks.ToHashSet());

        // The other half, pinned against the CODE rather than against the
        // contract's own consistency: no name the contract forbids a repair
        // for may appear in this app's repairable list.
        Assert.Empty(neverOffered.Intersect(SiteHealthFinding.RepairableChecks));
    }

    /// <summary>
    /// The one refusal the contract carries is worded here exactly as it is
    /// there -- <c>siteHealth.repair.refusedWhenSomethingIsInTheWay</c>. The
    /// sentence is not retyped: it is read from the contract, its placeholders
    /// filled, and compared with what this app says.
    /// </summary>
    [Fact]
    public void TheRefusalSentenceIsTheContractsWordForWord()
    {
        var refused = SiteHealth["repair"]!["refusedWhenSomethingIsInTheWay"]!;
        var cases = refused["cases"]!.AsArray();
        Assert.NotEmpty(cases);

        foreach (var refusal in cases)
        {
            // An unknown refusal case would have appeared with nothing behind it.
            Assert.Equal("sectionIndexMissing", refusal!["check"]!.ToString());
            Assert.Equal("refused", refusal["expect"]!.ToString());
            string sentence = refusal["sentence"]!.ToString()
                .Replace("{course}", "ICS3U")
                .Replace("{section}", "2");
            Assert.Equal(sentence, SiteHealthRepair.FolderWhereTheFrontPageBelongs("ICS3U", 2));
        }
    }

    /// <summary>
    /// Rule 1 again, for the refusal -- which the machinery sweep below does
    /// not reach, because it walks <c>checks</c> and this sentence lives under
    /// <c>repair</c>.
    /// </summary>
    [Fact]
    public void TheRefusalSentenceNamesNoMachinery()
    {
        string said = SiteHealthRepair.FolderWhereTheFrontPageBelongs("ICS3U", 1).ToLowerInvariant();
        foreach (string word in new[] { "toolchain", "script", "docker", "container", "wsl",
                                        "python", "stdout", "quartz", "repository", "config" })
            Assert.DoesNotContain(word, said);
    }

    /// <summary>
    /// The trail line carries the stable check NAME, never the product
    /// wording. <c>activityTrail.mustRecord</c> -> "folder problem found" says
    /// so, and the reason is that the sentence gets reworded while the name is
    /// what a person searching the trail can match against the contract.
    /// </summary>
    [Fact]
    public void TheTrailLineCarriesTheCheckNameAndNotTheWording()
    {
        var check = SiteHealth["checks"]!.AsArray()
            .First(c => c!["name"]!.ToString() == "noGradedFolders")!;
        string wording = check["sentence"]!.ToString();

        var finding = new SiteHealthFinding("noGradedFolders", wording, "d", false, "ICS3U", 1);

        Assert.Contains("noGradedFolders", finding.TrailSentence);
        Assert.DoesNotContain(wording, finding.TrailSentence);
    }

    // ---- Parsing the shape the launcher actually produces ---------------

    [Fact]
    public void OrdinaryBuildOutputIsNotAFinding()
    {
        Assert.Null(SiteHealthFinding.Parse("Building ICS3U Section 1..."));
        Assert.Null(SiteHealthFinding.Parse(""));
        Assert.Null(SiteHealthFinding.Parse(null));
    }

    [Fact]
    public void AMalformedHealthLineIsIgnoredRatherThanThrowing()
    {
        Assert.Null(SiteHealthFinding.Parse($"{SiteHealthFinding.Marker} not json at all"));
        Assert.Null(SiteHealthFinding.Parse($"{SiteHealthFinding.Marker} {{}}"));
        Assert.Null(SiteHealthFinding.Parse(SiteHealthFinding.Marker));
    }

    [Fact]
    public void AFindingIsStillFoundWhenGluedToOtherOutput()
    {
        string line = "  ...done. " + SiteHealth["marker"]!["examples"]!.AsArray()[0]!.ToString();
        Assert.NotNull(SiteHealthFinding.Parse(line));
    }

    [Fact]
    public void TheSameProblemFromTwoBuildsIsReportedOnce()
    {
        string example = SiteHealth["marker"]!["examples"]!.AsArray()[0]!.ToString();
        string other = SiteHealth["marker"]!["examples"]!.AsArray()[1]!.ToString();

        // Expected names come from the same two lines, not from a literal:
        // the ORDER of the contract's examples carries no product meaning, and
        // a test that goes red when somebody reorders them is a test nobody
        // trusts.
        var expected = new[]
        {
            SiteHealthFinding.Parse(example)!.Name,
            SiteHealthFinding.Parse(other)!.Name,
        };
        Assert.NotEqual(expected[0], expected[1]);

        var findings = SiteHealthFinding.FindingsIn(new[]
        {
            "Building...", example, "Building again...", example, other,
        });

        Assert.Equal(2, findings.Count);
        Assert.Equal(expected, findings.Select(f => f.Name).ToArray());
    }

    [Fact]
    public void TheSameCheckInTwoSectionsIsTwoProblems()
    {
        string One(int section) =>
            $"{SiteHealthFinding.Marker} {{\"name\": \"sectionIndexMissing\", \"sentence\": \"s\", " +
            $"\"detail\": \"d\", \"fixable\": true, \"course\": \"ICS3U\", \"section\": {section}}}";

        var findings = SiteHealthFinding.FindingsIn(new[] { One(1), One(2), One(1) });

        Assert.Equal(2, findings.Count);
        Assert.Equal(new[] { 1, 2 }, findings.Select(f => f.Section).ToArray());
    }
}

File: mac-app/Tests/QuartzTeachersTests/SiteHealthContractTests.swift
import XCTest
@testable import QuartzTeachers

/// The site-health contract, run against this app's own reading of it.
///
/// Without this the contract's `siteHealth` section was exercised only by the
/// Python that WRITES those lines — so the marker prefix and the wording were
/// unpinned across exactly the boundary the contract exists to protect. A
/// toolchain that changed its prefix would have gone on printing findings that
/// neither app could see, and nothing would have failed.
@MainActor
final class SiteHealthContractTests: XCTestCase {

    // MARK: - Stored properties

    private var siteHealth: [String: Any] = [:]

    // MARK: - Functions

    override func setUpWithError() throws {
        let url: URL = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent().deletingLastPathComponent()
            .deletingLastPathComponent().deletingLastPathComponent()
            .appendingPathComponent("contracts/shared-rules.json")
        let all: [String: Any] = try XCTUnwrap(
            try JSONSerialization.jsonObject(with: try Data(contentsOf: url)) as? [String: Any]
        )
        siteHealth = try XCTUnwrap(all["siteHealth"] as? [String: Any])
    }

    func testTheMarkerPrefixMatchesTheContract() throws {
        let marker: [String: Any] = try XCTUnwrap(siteHealth["marker"] as? [String: Any])
        let prefix: String = try XCTUnwrap(marker["prefix"] as? String)
        XCTAssertEqual(SiteHealthFinding.markerPrefix, prefix,
                       "the app would stop seeing findings the toolchain still prints")
    }

    /// Every sentence in the contract must survive the round trip the real
    /// path takes — printed as a marker line, read back by the app — with its
    /// placeholders filled in and nothing left showing.
    func testEveryContractSentenceSurvivesTheRoundTrip() throws {
        let checks: [[String: Any]] = try XCTUnwrap(siteHealth["checks"] as? [[String: Any]])
        XCTAssertGreaterThanOrEqual(checks.count, 5, "the contract lost checks")

        for check in checks {
            let name: String = try XCTUnwrap(check["name"] as? String)
            let sentence: String = try XCTUnwrap(check["sentence"] as? String)
                .replacingOccurrences(of: "{course}", with: "ICS3U")
                .replacingOccurrences(of: "{section}", with: "1")
            let detail: String = try XCTUnwrap(check["detail"] as? String)
            let fixable: Bool = try XCTUnwrap(check["fixable"] as? Bool)

            let payload: [String: Any] = [
                "name": name, "sentence": sentence, "detail": detail,
                "fixable": fixable, "course": "ICS3U", "section": 1,
            ]
            let data: Data = try JSONSerialization.data(withJSONObject: payload)
            let line: String = SiteHealthFinding.markerPrefix + " "
                + (String(data: data, encoding: .utf8) ?? "")

            let found: [SiteHealthFinding] = SiteHealthFinding.findings(in: line)
            XCTAssertEqual(found.count, 1, name)
            XCTAssertEqual(found.first?.name, name)
            XCTAssertEqual(found.first?.sentence, sentence, name)
            XCTAssertEqual(found.first?.fixable, fixable, name)
        }
    }

    /// Parses lines the TOOLCHAIN actually printed.
    ///
    /// The round-trip test above builds its own JSON, so it proves the app can
    /// read the app — it would not catch `section` emitted as a string, a
    /// renamed key, or a changed separator. These examples are captured from
    /// `site_health.announce` and re-checked against it by
    /// `scripts/test_site_health.py`, so the two halves cannot drift apart
    /// without one of the two suites failing.
    func testRealToolchainOutputIsUnderstood() throws {
        let marker: [String: Any] = try XCTUnwrap(siteHealth["marker"] as? [String: Any])
        let examples: [String] = try XCTUnwrap(marker["examples"] as? [String])
        XCTAssertFalse(examples.isEmpty, "the contract carries no example output")

        for example in examples {
            let found: [SiteHealthFinding] = SiteHealthFinding.findings(in: example)
            XCTAssertEqual(found.count, 1, example)
            let finding: SiteHealthFinding = try XCTUnwrap(found.first)
            XCTAssertFalse(finding.name.isEmpty)
            XCTAssertFalse(finding.sentence.isEmpty)
            XCTAssertFalse(finding.detail.isEmpty, "the dialog shows the detail too")
            XCTAssertEqual(finding.course, "ICS3U")
            XCTAssertEqual(finding.section, 1,
                           "section must arrive as a number, not fall back to 0")
        }
    }

    /// The repair lists in the contract must match what the app will actually
    /// offer.
    ///
    /// Nothing read `siteHealth.repair` when it was added, so the lists could
    /// have drifted from `canRepair` silently — a contract section that is
    /// documentation rather than a gate. The `neverOffered` half is the one
    /// that matters: those are the checks a fix would SATISFY without fixing.
    func testTheRepairListsMatchWhatTheAppWillOffer() throws {
        let repair: [String: Any] = try XCTUnwrap(siteHealth["repair"] as? [String: Any])

        let offered: [String] = try XCTUnwrap(
            (repair["offered"] as? [String: Any])?["checks"] as? [String]
        )
        let neverOffered: [String] = try XCTUnwrap(
            (repair["neverOffered"] as? [String: Any])?["checks"] as? [String]
        )

        for name in offered {
            XCTAssertTrue(
                SiteHealthRepair.canRepair(sample(named: name, fixable: true)),
                "\(name) is listed as repairable and the app will not offer it"
            )
        }
        for name in neverOffered {
            XCTAssertFalse(
                SiteHealthRepair.canRepair(sample(named: name, fixable: true)),
                "\(name) must never be offered a fix, even when the toolchain "
                + "marks it fixable — a fix that satisfies the check without "
                + "restoring the feature is worse than none"
            )
        }

        // And between them they must account for every check, or a new one
        // silently belongs to neither list.
        let checks: [[String: Any]] = try XCTUnwrap(siteHealth["checks"] as? [[String: Any]])
        var everyName: [String] = []
        for check in checks {
            everyName.append(try XCTUnwrap(check["name"] as? String))
        }
        XCTAssertEqual(
            Set(offered).union(Set(neverOffered)), Set(everyName),
            "every check must be in exactly one of the repair lists"
        )
    }

    /// The sentence a teacher reads when a FOLDER is sitting where a section's
    /// front page belongs, deserialised rather than retyped.
    ///
    /// Both apps have to say one thing about one problem. Windows answered
    /// `Failed` here until 2026-09-07, with the generic "check the folder isn't
    /// locked or read-only", which sent a teacher to look at permissions on a
    /// folder that is not locked — so this is the sentence it adopted, and both
    /// apps say it now.
    func testTheRefusalSentenceIsTheOneInTheContract() throws {
        let repair: [String: Any] = try XCTUnwrap(siteHealth["repair"] as? [String: Any])
        let refused: [String: Any] = try XCTUnwrap(
            repair["refusedWhenSomethingIsInTheWay"] as? [String: Any]
        )
        let cases: [[String: Any]] = try XCTUnwrap(refused["cases"] as? [[String: Any]])
        XCTAssertFalse(cases.isEmpty, "the contract carries no refusal case")

        for refusal in cases {
            XCTAssertEqual(refusal["check"] as? String, "sectionIndexMissing",
                           "an unknown refusal case has appeared with nothing behind it")
            XCTAssertEqual(refusal["expect"] as? String, "refused")
            let sentence: String = try XCTUnwrap(refusal["sentence"] as? String)
                .replacingOccurrences(of: "{course}", with: "ICS3U")
                .replacingOccurrences(of: "{section}", with: "2")
            XCTAssertEqual(
                SiteHealthRepair.folderWhereTheFrontPageBelongs(course: "ICS3U", section: 2),
                sentence,
                "the app and the contract word the same refusal differently"
            )
        }
    }

    /// Rule 1 again, for the refusal — which `testNoCheckNamesTheMachinery`
    /// below does not reach, because it walks `checks` and this sentence lives
    /// under `repair`.
    func testTheRefusalSentenceNamesNoMachinery() throws {
        let said: String = SiteHealthRepair
            .folderWhereTheFrontPageBelongs(course: "ICS3U", section: 1)
            .lowercased()
        for word in ["toolchain", "script", "docker", "container", "wsl", "python",
                     "stdout", "quartz", "repository", "config"] {
            XCTAssertFalse(said.contains(word), "the refusal says \"\(word)\" to a teacher")
        }
    }

    private func sample(named name: String, fixable: Bool) -> SiteHealthFinding {
        return SiteHealthFinding(
            name: name, sentence: "s", detail: "d",
            fixable: fixable, course: "ICS3U", section: 1
        )
    }

    /// Rule 1: the interface never names the machinery. These sentences are
    /// shown to a teacher verbatim — in a dialog, and in the assistant's
    /// answer — so a stray "container" or "script" would reach them directly.
    func testNoCheckNamesTheMachinery() throws {
        let checks: [[String: Any]] = try XCTUnwrap(siteHealth["checks"] as? [[String: Any]])
        let forbidden: [String] = [
            "toolchain", "script", "docker", "container", "wsl", "python",
            "json", "stdout", "quartz", "repository", "config",
        ]
        for check in checks {
            let name: String = (check["name"] as? String) ?? "unnamed"
            let shown: String = ((check["sentence"] as? String) ?? "")
                + " " + ((check["detail"] as? String) ?? "")
            for word in forbidden {
                XCTAssertFalse(
                    shown.lowercased().contains(word),
                    "\(name) says \"\(word)\" to a teacher"
                )
            }
        }
    }

    /// The rule that must not be quietly softened later: a scheduled deploy
    /// publishes anyway. A slightly inaccurate map is a paper cut; a site
    /// update a teacher was counting on that silently did not happen is not.
    func testTheScheduledDeployRuleIsStillWrittenDown() throws {
        let rule: String = try XCTUnwrap(siteHealth["scheduledDeployPublishesAnyway"] as? String)
        XCTAssertTrue(rule.lowercased().contains("never refuses"), rule)
    }
}
